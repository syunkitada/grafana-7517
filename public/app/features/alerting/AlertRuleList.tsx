import React, { PureComponent } from 'react';
import { hot } from 'react-hot-loader';
import { connect } from 'react-redux';
import classNames from 'classnames';
import PageHeader from 'app/core/components/PageHeader/PageHeader';
import appEvents from 'app/core/app_events';
import Highlighter from 'react-highlight-words';
import { updateLocation } from 'app/core/actions';
import { getNavModel } from 'app/core/selectors/navModel';
import { NavModel, StoreState, AlertRule } from 'app/types';
import { getAlertRulesAsync } from './state/actions';

interface Props {
  navModel: NavModel;
  alertRules: AlertRule[];
  updateLocation: typeof updateLocation;
  getAlertRulesAsync: typeof getAlertRulesAsync;
}

interface State {
  rules: AlertRule[];
  search: string;
  stateFilter: string;
}

export class AlertRuleList extends PureComponent<Props, State> {
  stateFilters = [
    { text: 'All', value: 'all' },
    { text: 'OK', value: 'ok' },
    { text: 'Not OK', value: 'not_ok' },
    { text: 'Alerting', value: 'alerting' },
    { text: 'No Data', value: 'no_data' },
    { text: 'Paused', value: 'paused' },
  ];

  constructor(props) {
    super(props);

    this.state = {
      rules: [],
      search: '',
      stateFilter: '',
    };
  }

  componentDidMount() {
    this.fetchRules();
  }

  onStateFilterChanged = evt => {
    this.props.updateLocation({
      query: { state: evt.target.value },
    });
    this.fetchRules();
  };

  async fetchRules() {
    await this.props.getAlertRulesAsync();

    // this.props.alertList.loadRules({
    //   state: this.props.view.query.get('state') || 'all',
    // });
  }

  onOpenHowTo = () => {
    appEvents.emit('show-modal', {
      src: 'public/app/features/alerting/partials/alert_howto.html',
      modalClass: 'confirm-modal',
      model: {},
    });
  };

  onSearchQueryChange = evt => {
    // this.props.alertList.setSearchQuery(evt.target.value);
  };

  render() {
    const { navModel, alertRules } = this.props;
    const { search, stateFilter } = this.state;

    return (
      <div>
        <PageHeader model={navModel} />
        <div className="page-container page-body">
          <div className="page-action-bar">
            <div className="gf-form gf-form--grow">
              <label className="gf-form--has-input-icon gf-form--grow">
                <input
                  type="text"
                  className="gf-form-input"
                  placeholder="Search alerts"
                  value={search}
                  onChange={this.onSearchQueryChange}
                />
                <i className="gf-form-input-icon fa fa-search" />
              </label>
            </div>
            <div className="gf-form">
              <label className="gf-form-label">States</label>

              <div className="gf-form-select-wrapper width-13">
                <select className="gf-form-input" onChange={this.onStateFilterChanged} value={stateFilter}>
                  {this.stateFilters.map(AlertStateFilterOption)}
                </select>
              </div>
            </div>

            <div className="page-action-bar__spacer" />

            <a className="btn btn-secondary" onClick={this.onOpenHowTo}>
              <i className="fa fa-info-circle" /> How to add an alert
            </a>
          </div>

          <section>
            <ol className="alert-rule-list">
              {alertRules.map(rule => <AlertRuleItem rule={rule} key={rule.id} search={search} />)}
            </ol>
          </section>
        </div>
      </div>
    );
  }
}

function AlertStateFilterOption({ text, value }) {
  return (
    <option key={value} value={value}>
      {text}
    </option>
  );
}

export interface AlertRuleItemProps {
  rule: AlertRule;
  search: string;
}

export class AlertRuleItem extends React.Component<AlertRuleItemProps, any> {
  toggleState = () => {
    // this.props.rule.togglePaused();
  };

  renderText(text: string) {
    return (
      <Highlighter
        highlightClassName="highlight-search-match"
        textToHighlight={text}
        searchWords={[this.props.search]}
      />
    );
  }

  render() {
    const { rule } = this.props;

    const stateClass = classNames({
      fa: true,
      'fa-play': rule.state === 'paused',
      'fa-pause': rule.state !== 'paused',
    });

    const ruleUrl = `${rule.url}?panelId=${rule.panelId}&fullscreen=true&edit=true&tab=alert`;

    return (
      <li className="alert-rule-item">
        <span className={`alert-rule-item__icon ${rule.stateClass}`}>
          <i className={rule.stateIcon} />
        </span>
        <div className="alert-rule-item__body">
          <div className="alert-rule-item__header">
            <div className="alert-rule-item__name">
              <a href={ruleUrl}>{this.renderText(rule.name)}</a>
            </div>
            <div className="alert-rule-item__text">
              <span className={`${rule.stateClass}`}>{this.renderText(rule.stateText)}</span>
              <span className="alert-rule-item__time"> for {rule.stateAge}</span>
            </div>
          </div>
          {rule.info && <div className="small muted alert-rule-item__info">{this.renderText(rule.info)}</div>}
        </div>

        <div className="alert-rule-item__actions">
          <button
            className="btn btn-small btn-inverse alert-list__btn width-2"
            title="Pausing an alert rule prevents it from executing"
            onClick={this.toggleState}
          >
            <i className={stateClass} />
          </button>
          <a className="btn btn-small btn-inverse alert-list__btn width-2" href={ruleUrl} title="Edit alert rule">
            <i className="icon-gf icon-gf-settings" />
          </a>
        </div>
      </li>
    );
  }
}

const mapStateToProps = (state: StoreState) => ({
  navModel: getNavModel(state.navIndex, 'alert-list'),
  alertRules: state.alertRules,
});

const mapDispatchToProps = {
  updateLocation,
  getAlertRulesAsync,
};

export default hot(module)(connect(mapStateToProps, mapDispatchToProps)(AlertRuleList));
