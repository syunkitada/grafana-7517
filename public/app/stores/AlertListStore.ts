import { types, getEnv, flow } from 'mobx-state-tree';
import moment from 'moment';
import alertDef from 'app/features/alerting/alert_def';

function setStateFields(rule, state) {
  let stateModel = alertDef.getStateDisplayModel(state);
  rule.state = state;
  rule.stateText = stateModel.text;
  rule.stateIcon = stateModel.iconClass;
  rule.stateClass = stateModel.stateClass;
  rule.stateAge = moment(rule.newStateDate)
    .fromNow()
    .replace(' ago', '');
}

export const AlertRule = types
  .model('AlertRule', {
    id: types.identifier(types.number),
    dashboardId: types.number,
    panelId: types.number,
    name: types.string,
    state: types.string,
    stateText: types.string,
    stateIcon: types.string,
    stateClass: types.string,
    stateAge: types.string,
    info: types.optional(types.string, ''),
    dashboardUri: types.string,
  })
  .views(self => ({
    get isPaused() {
      return self.state === 'paused';
    },
  }))
  .actions(self => ({
    /**
     * will toggle alert rule paused state
     */
    togglePaused: flow(function* togglePaused() {
      let backendSrv = getEnv(self).backendSrv;

      var payload = { paused: self.isPaused };
      let res = yield backendSrv.post(`/api/alerts/${self.id}/pause`, payload);
      setStateFields(self, res.state);
      self.info = '';
    }),
  }));

type IAlertRuleType = typeof AlertRule.Type;
export interface IAlertRule extends IAlertRuleType {}

export const AlertListStore = types
  .model('AlertListStore', {
    rules: types.array(AlertRule),
    stateFilter: types.optional(types.string, 'all'),
  })
  .actions(self => ({
    loadRules: flow(function* load(filters) {
      let backendSrv = getEnv(self).backendSrv;

      // store state filter used in api query
      self.stateFilter = filters.state;

      let apiRules = yield backendSrv.get('/api/alerts', filters);

      self.rules.clear();

      for (let rule of apiRules) {
        setStateFields(rule, rule.state);

        if (rule.executionError) {
          rule.info = 'Execution Error: ' + rule.executionError;
        }

        if (rule.evalData && rule.evalData.noData) {
          rule.info = 'Query returned no data';
        }

        self.rules.push(AlertRule.create(rule));
      }
    }),
  }));
