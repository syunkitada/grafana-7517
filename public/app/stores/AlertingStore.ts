import { types, getEnv, flow } from 'mobx-state-tree';
import moment from 'moment';
import alertDef from 'app/features/alerting/alert_def';

export const AlertRule = types.model('AlertRule', {
  id: types.identifier(types.number),
  dashboardId: types.number,
  panelId: types.number,
  name: types.string,
  state: types.string,
  stateText: types.string,
  stateIcon: types.string,
  stateClass: types.string,
  stateAge: types.string,
  info: types.optional(types.string, ''),
  dashboardUri: types.string,
});

export const AlertingStore = types
  .model('AlertingStore', {
    rules: types.array(AlertRule),
  })
  .actions(self => ({
    loadRules: flow(function* load() {
      let backendSrv = getEnv(self).backendSrv;

      let rules = yield backendSrv.get('/api/alerts');

      self.rules.clear();

      for (let rule of rules) {
        let stateModel = alertDef.getStateDisplayModel(rule.state);
        rule.stateText = stateModel.text;
        rule.stateIcon = stateModel.iconClass;
        rule.stateClass = stateModel.stateClass;
        rule.stateAge = moment(rule.newStateDate)
          .fromNow()
          .replace(' ago', '');

        if (rule.executionError) {
          rule.info = 'Execution Error: ' + rule.executionError;
        }

        if (rule.evalData && rule.evalData.noData) {
          rule.info = 'Query returned no data';
        }

        self.rules.push(AlertRule.create(rule));
      }
    }),
  }));
