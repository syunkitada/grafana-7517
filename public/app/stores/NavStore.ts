import { types, getEnv } from 'mobx-state-tree';
import _ from 'lodash';

export const NavItem = types.model('NavItem', {
  id: types.identifier(types.string),
  text: types.string,
  url: types.optional(types.string, ''),
  subTitle: types.optional(types.string, ''),
  icon: types.optional(types.string, ''),
  img: types.optional(types.string, ''),
  active: types.optional(types.boolean, false),
  children: types.optional(types.array(types.late(() => NavItem)), []),
});

export const NavStore = types
  .model('NavStore', {
    main: types.maybe(NavItem),
    node: types.maybe(NavItem),
  })
  .actions(self => ({
    load(...args) {
      var children = getEnv(self).navTree;
      let main, node;
      let parents = [];

      for (let id of args) {
        node = _.find(children, { id: id });
        if (!node) {
          throw new Error(`NavItem with id ${id} not found`);
        }

        children = node.children;
        parents.push(node);
      }

      main = parents[parents.length - 2];

      if (main.children) {
        for (let item of main.children) {
          item.active = false;

          if (item.url === node.url) {
            item.active = true;
          }
        }
      }

      self.main = NavItem.create(main);
      self.node = NavItem.create(node);
    },
  }));
