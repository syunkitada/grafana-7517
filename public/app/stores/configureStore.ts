import { createStore, applyMiddleware, compose, combineReducers } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import sharedReducers from 'app/core/reducers';

const rootReducer = combineReducers({
  ...sharedReducers
});

export let store;

export function configureStore() {
  const composeEnhancers = (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
  store = createStore(rootReducer, {}, composeEnhancers(applyMiddleware(thunk, createLogger())));
}
