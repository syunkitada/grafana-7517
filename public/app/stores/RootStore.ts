import { types } from 'mobx-state-tree';
import { SearchStore } from './SearchStore';
import { ServerStatsStore } from './ServerStatsStore';
import { NavStore } from './NavStore/NavStore';

export const RootStore = types.model({
  search: types.optional(SearchStore, {
    sections: [],
  }),
  serverStats: types.optional(ServerStatsStore, {
    stats: [],
  }),
  nav: types.optional(NavStore, {}),
});

type IRootStoreType = typeof RootStore.Type;
export interface IRootStore extends IRootStoreType {}
