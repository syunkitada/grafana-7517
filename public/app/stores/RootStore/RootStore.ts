import { types } from 'mobx-state-tree';
import { NavStore } from './../NavStore/NavStore';
import { ViewStore } from './../ViewStore/ViewStore';
import { PermissionsStore } from './../PermissionsStore/PermissionsStore';

export const RootStore = types.model({
  nav: types.optional(NavStore, {}),
  permissions: types.optional(PermissionsStore, {
    fetching: false,
    items: [],
  }),
  view: types.optional(ViewStore, {
    path: '',
    query: {},
    routeParams: {},
  }),
});

type RootStoreType = typeof RootStore.Type;
export interface RootStoreInterface extends RootStoreType {}
