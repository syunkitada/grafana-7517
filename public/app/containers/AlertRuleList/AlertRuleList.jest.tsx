import React from 'react';
// import renderer from 'react-test-renderer';
import moment from 'moment';
import { AlertRuleList, AlertRuleItem } from './AlertRuleList';
import { RootStore } from 'app/stores/RootStore';
import { backendSrv, createNavTree } from 'test/mocks/common';
import { shallow } from 'enzyme';

describe('AlertRuleList', () => {
  let page, store;

  beforeAll(done => {
    backendSrv.get.mockReturnValue(
      Promise.resolve([
        {
          id: 11,
          dashboardId: 58,
          panelId: 3,
          name: 'Panel Title alert',
          state: 'ok',
          newStateDate: moment()
            .subtract(5, 'minutes')
            .format(),
          evalData: {},
          executionError: '',
          dashboardUri: 'db/mygool',
        },
      ])
    );

    store = RootStore.create(
      {},
      {
        backendSrv: backendSrv,
        navTree: createNavTree('alerting', 'alert-list'),
      }
    );

    page = shallow(<AlertRuleList store={store} />)
      .first()
      .shallow();
    setTimeout(done, 100);
    //page = renderer.create(<AlertRuleList store={store} />);
  });

  it('should call api to get rules', () => {
    expect(backendSrv.get.mock.calls[0][0]).toEqual('/api/alerts');
  });

  it('should render 1 rule', () => {
    console.log(page.find('.card-section').debug());
    expect(page.find(AlertRuleItem)).toHaveLength(1);
  });
});
