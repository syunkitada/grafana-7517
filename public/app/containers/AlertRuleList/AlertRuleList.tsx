import React from 'react';
import classNames from 'classnames';
import { inject, observer } from 'mobx-react';
import PageHeader from 'app/core/components/PageHeader/PageHeader';
import { IRootStore } from 'app/stores/RootStore';
import { IAlertRule } from 'app/stores/AlertListStore';
import appEvents from 'app/core/app_events';

export interface AlertRuleListProps {
  store: IRootStore;
}

@inject('store')
@observer
export class AlertRuleList extends React.Component<AlertRuleListProps, any> {
  stateFilters = [
    { text: 'All', value: 'all' },
    { text: 'OK', value: 'ok' },
    { text: 'Not OK', value: 'not_ok' },
    { text: 'Alerting', value: 'alerting' },
    { text: 'No Data', value: 'no_data' },
    { text: 'Paused', value: 'paused' },
  ];

  constructor(props) {
    super(props);

    const store = this.props.store;

    store.nav.load('alerting', 'alert-list');
    store.alertList.setStateFilter(store.view.query.get('state') || 'all');
    store.alertList.loadRules();
  }

  onStateFilterChanged = evt => {
    this.props.store.alertList.setStateFilter(evt.target.value);
    this.props.store.view.updateQuery({ state: evt.target.value });
    this.props.store.alertList.loadRules();
  };

  onOpenHowTo = () => {
    appEvents.emit('show-modal', {
      src: 'public/app/features/alerting/partials/alert_howto.html',
      modalClass: 'confirm-modal',
      model: {},
    });
  };

  render() {
    const { nav, alertList } = this.props.store;

    return (
      <div>
        <PageHeader model={nav as any} />
        <div className="page-container page-body">
          <div className="page-action-bar">
            <div className="gf-form">
              <label className="gf-form-label">Filter by state</label>

              <div className="gf-form-select-wrapper width-13">
                <select className="gf-form-input" onChange={this.onStateFilterChanged} value={alertList.stateFilter}>
                  {this.stateFilters.map(AlertStateFilterOption)}
                </select>
              </div>
            </div>

            <div className="page-action-bar__spacer" />

            <a className="btn btn-secondary" onClick={this.onOpenHowTo}>
              <i className="fa fa-info-circle" /> How to add an alert
            </a>
          </div>

          <section className="card-section card-list-layout-list">
            <ol className="card-list">{alertList.rules.map(rule => <AlertRuleItem rule={rule} key={rule.id} />)}</ol>
          </section>
        </div>
      </div>
    );
  }
}

function AlertStateFilterOption({ text, value }) {
  return (
    <option key={value} value={value}>
      {text}
    </option>
  );
}

export interface AlertRuleItemProps {
  rule: IAlertRule;
}

@observer
export class AlertRuleItem extends React.Component<AlertRuleItemProps, any> {
  toggleState = () => {
    this.props.rule.togglePaused();
  };

  render() {
    const { rule } = this.props;

    let stateClass = classNames({
      fa: true,
      'fa-play': rule.isPaused,
      'fa-pause': !rule.isPaused,
    });

    let ruleUrl = `dashboard/${rule.dashboardUri}?panelId=${rule.panelId}&fullscreen&edit&tab=alert`;

    return (
      <li className="card-item-wrapper">
        <div className="card-item card-item--alert">
          <div className="card-item-header">
            <div className="card-item-type">
              <a
                className="card-item-cog"
                title="Pausing an alert rule prevents it from executing"
                onClick={this.toggleState}
              >
                <i className={stateClass} />
              </a>
              <a className="card-item-cog" href={ruleUrl} title="Edit alert rule">
                <i className="icon-gf icon-gf-settings" />
              </a>
            </div>
          </div>
          <div className="card-item-body">
            <div className="card-item-details">
              <div className="card-item-name">
                <a href={ruleUrl}>{rule.name}</a>
              </div>
              <div className="card-item-sub-name">
                <span className={`alert-list-item-state ${rule.stateClass}`}>
                  <i className={rule.stateIcon} /> {rule.stateText}
                </span>
                <span> for {rule.stateAge}</span>
              </div>
              {rule.info && <div className="small muted">{rule.info}</div>}
            </div>
          </div>
        </div>
      </li>
    );
  }
}
