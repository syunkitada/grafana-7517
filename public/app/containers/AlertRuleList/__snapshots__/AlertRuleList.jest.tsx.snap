// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlertRuleList should render 1 rule 1`] = `
<li
  className="card-item-wrapper"
>
  <div
    className="card-item card-item--alert"
  >
    <div
      className="card-item-header"
    >
      <div
        className="card-item-type"
      >
        <a
          className="card-item-cog"
          onClick={[Function]}
          title="Pausing an alert rule prevents it from executing"
        >
          <i
            className="fa fa-pause"
          />
        </a>
        <a
          className="card-item-cog"
          href="dashboard/db/mygool?panelId=3&fullscreen&edit&tab=alert"
          title="Edit alert rule"
        >
          <i
            className="icon-gf icon-gf-settings"
          />
        </a>
      </div>
    </div>
    <div
      className="card-item-body"
    >
      <div
        className="card-item-details"
      >
        <div
          className="card-item-name"
        >
          <a
            href="dashboard/db/mygool?panelId=3&fullscreen&edit&tab=alert"
          >
            Panel Title alert
          </a>
        </div>
        <div
          className="card-item-sub-name"
        >
          <span
            className="alert-list-item-state alert-state-ok"
          >
            <i
              className="icon-gf icon-gf-online"
            />
             
            OK
          </span>
          <span>
             for 
            5 minutes
          </span>
        </div>
      </div>
    </div>
  </div>
</li>
`;
